File: Cplusplus/add_two_numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        //Create a dummy head so as to return its next node as result
        ListNode *dummy_head = new ListNode();
        ListNode *result = dummy_head;
        int sum=0;
        //Run loop till either l1 or l2 present or carry value in sum present
        while(l1||l2||sum){
            sum +=(l1 ? l1->val : 0) + (l2? l2->val : 0);
            result->next =new ListNode(sum%10);
            result = result ->next;
            sum/=10; //Save carry
            
            if(l1){l1=l1->next;}
            if(l2){l2=l2->next;}
        }
        return dummy_head->next;
    }
};

File: Python/add_two_numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        """
        Questions: Are they of same size? 
        What's min & max size?(Is it 0 to 2^32?)
        Can they be empty?
        """
        #Create a dummy head so as to return its next node as result
        result_head = ListNode(0) 
        result = result_head
        res_val =0
        #Run loop till either l1 and l2 end. In that case, run loop to create node using res_val(carry)
        while l1 or l2 or res_val: 
            val1 = l1.val if l1!=None else 0
            val2 = l2.val if l2!=None else 0
            
            res_val+= val1 + val2
            result.next = ListNode(res_val%10)
            result = result.next
            res_val =res_val//10
            
            if l1:
                l1=l1.next
            if l2:
                l2= l2.next
                
        return result_head.next         
